database/CoreDatabase.java

```diff
@@ -494,39 +492,39 @@ public void delete() {
 
     public static class IsolationManager {
 
-        private final ConcurrentMap<CoreTransaction.Data, CommitState> commitStates;
-        private final ConcurrentNavigableMap<Long, Set<CoreTransaction.Data>> commitTimeline;
+        private final ConcurrentSet<CoreTransaction.Data> uncommitted;
+        private final ConcurrentSet<CoreTransaction.Data> committing;
+        private final ConcurrentSet<CoreTransaction.Data> committed;
         private final AtomicBoolean cleanupRunning;
 
-        private enum CommitState {UNCOMMITTED, COMMITTING, COMMITTED}
-
         IsolationManager() {
-            this.cleanupRunning = new AtomicBoolean(false);
-            this.commitStates = new ConcurrentHashMap<>();
-            this.commitTimeline = new ConcurrentSkipListMap<>();
+            uncommitted = new ConcurrentSet<>();
+            committing = new ConcurrentSet<>();
+            committed = new ConcurrentSet<>();
+            cleanupRunning = new AtomicBoolean(false);
         }
 
         void opened(CoreTransaction.Data transaction) {
-            commitStates.put(transaction, CommitState.UNCOMMITTED);
+            uncommitted.add(transaction);
         }
 
         public Set<CoreTransaction.Data> validateOverlappingAndStartCommit(CoreTransaction.Data txn) {
             Set<CoreTransaction.Data> transactions;
             synchronized (this) {
                 transactions = commitMayConflict(txn);
                 transactions.forEach(other -> validateIsolation(txn, other));
-                commitStates.put(txn, CommitState.COMMITTING);
+                committing.add(txn);
+                uncommitted.remove(txn);
             }
             return transactions;
         }
 
         private Set<CoreTransaction.Data> commitMayConflict(CoreTransaction.Data txn) {
             if (!txn.dataStorage.hasTrackedWrite()) return set();
-            return iterate(commitStates.entrySet())
-                    .filter(e -> e.getValue() == CommitState.COMMITTING ||
-                            (e.getValue() == CommitState.COMMITTED && e.getKey().snapshotEnd().get() > txn.snapshotStart()))
-                    .map(Map.Entry::getKey)
-                    .toSet();
+            Set<CoreTransaction.Data> mayConflict = new HashSet<>(committing);
+            iterate(committed).filter(committed -> committed.snapshotEnd().get() > txn.snapshotStart())
+                    .forEachRemaining(mayConflict::add);
```

Comment:
this should be faster than the old way because we don't have to filter out all the uncommitted transactions at all, we only need to actually check the committed transaction age in the filter